lib/plus_plus/base.rb

```diff
@@ -1,73 +1,133 @@
+      EOQ
+    end
+
     def plus_plus(*args)
       options = args.extract_options!
+      plus_plus_warning_about_removal_update_method_from_options(options)
+
       association, column = args
 
-      self.after_create {
-        self.plus_plus_on_create_or_destroy association, column, options
-      }
+      after_create do
+        plus_plus_on_create_or_destroy(association, column, options)
+      end
 
-      self.after_destroy {
-        self.plus_plus_on_create_or_destroy association, column, options
-      }
+      after_destroy do
+        plus_plus_on_create_or_destroy(association, column, options)
+      end
     end
 
     def plus_plus_on_change(*args)
       options = args.extract_options!
+      plus_plus_warning_about_removal_update_method_from_options(options)
+
       association, column = args
 
-      self.after_update do
-        raise "No :changed option specified" if options[:changed].nil?
-        raise "No :plus option specified" if options[:plus].nil?
-        raise "No :minus option specified" if options[:minus].nil?
-        return unless self.changes.include?(options[:changed])
+      after_update do
+        raise 'No :changed option specified' if options[:changed].nil?
+        raise 'No :plus option specified' if options[:plus].nil?
+        raise 'No :minus option specified' if options[:minus].nil?
+
+        return unless changes.include?(options[:changed])
+
+        # Create a 'snapshot' of what the model did look like
+        dup = self.dup
+        changes.each { |k, v| dup[k] = v.first }
 
-        dup     = self.dup
         changed = options[:changed]
-        offset  = if options[:value]
-          options[:value].respond_to?(:call) ? self.instance_exec(&options[:value]) : options[:value]
-        else
-          1
-        end
+        prev_satisfied_for_minus = if options[:minus].respond_to?(:call)
+                                     dup.instance_exec(&options[:minus])
+                                   else
+                                     dup.send(changed) == options[:minus]
+                                   end
+        self_satisfied_for_plus = if options[:plus].respond_to?(:call)
+                                    instance_exec(&options[:plus])
+                                  else
+                                    send(changed) == options[:plus]
+                                  end
+        self_satisfied_for_minus = if options[:minus].respond_to?(:call)
+                                     instance_exec(&options[:minus])
+                                   else
+                                     send(changed) == options[:minus]
+                                   end
+        prev_satisfied_for_plus = if options[:plus].respond_to?(:call)
+                                    dup.instance_exec(&options[:plus])
+                                  else
+                                    dup.send(changed) == options[:plus]
+                                  end
```

Comment:
Seems like all these could be DRYed up into a method that takes an object (self or dup) and a potentially callable object (options[:plus] or options[:minus]).